pgjdbc/src/main/java/org/postgresql/core/v3/QueryExecutorImpl.java

```diff
@@ -2063,29 +2102,154 @@ private void processDeadPortals() throws IOException {
     // from there.
     boolean doneAfterRowDescNoData = false;
 
-    while (!endQuery) {
-      c = pgStream.receiveChar();
+    ProcessState(ResultHandler handler, int flags, boolean adaptiveFetch, @Nullable SQLThrowingRunnable onFinishedContext, @Nullable Consumer<IOException> onIOError) {
+      this.noResults = (flags & QueryExecutor.QUERY_NO_RESULTS) != 0;
+      this.bothRowsAndStatus = (flags & QueryExecutor.QUERY_BOTH_ROWS_AND_STATUS) != 0;
+      this.streamRows = (flags & QueryExecutor.QUERY_STREAM_ROWS) != 0;
+
+      assert (onFinishedContext == null) == (onIOError == null) : "Either both or neither of onFinishedContext and onIOError must be given";
+
+      this.adaptiveFetch = adaptiveFetch;
+      this.handler = handler;
+      this.onFinishedContext = onFinishedContext;
+      this.onIOError = onIOError;
+    }
+  }
+
+  /**
+   * Processes protocol events into results.
+   *
+   * @param handler The handler to feed the status and results
+   * @param flags Query flags
+   * @param adaptiveFetch state of adaptiveFetch to use during execution
+   * @param onFinished Callback that is invoked when streming query processing ends, can be null if
+   *     flags does not have {@link QueryExecutor#QUERY_STREAM_ROWS}
+   * @param onIOError Callback that is invoked when an IOExceptio occurs while processing streaming
+   *     query, can be null if flags does not have {@link QueryExecutor#QUERY_STREAM_ROWS}
+   * @return True if the results were handled synchronously
+   * @throws IOException if IO error occurs during synchronous processing
+   */
+  protected boolean processResults(ResultHandler handler, int flags, boolean adaptiveFetch,
+      @Nullable SQLThrowingRunnable onFinished, @Nullable Consumer<IOException> onIOError)
+      throws IOException {
+
+    if (streamingState != null) {
+      throw new IOException("Previous result is still streaming");
+    }
+    return processResultsImpl(handler, new ProcessState(handler, flags, adaptiveFetch, onFinished, onIOError)) == null;
+  }
+
+  private boolean waitForProtocolFree() throws SQLException {
+    ProcessState localStreamingState = streamingState;
+    if (localStreamingState == null || localStreamingState.reader == Thread.currentThread()) {
+      return true;
+    }
+    while (localStreamingState.reader != null) {
+      try {
+        this.wait();
+        localStreamingState = streamingState;
+        if (localStreamingState == null) {
+          return false;
+        }
+      } catch (InterruptedException ie) {
+        Thread.currentThread().interrupt();
+        throw new PSQLException(
+            GT.tr("Interrupted while waiting to obtain lock on database connection"),
+            PSQLState.OBJECT_NOT_IN_STATE, ie);
+      }
+    }
+    localStreamingState.reader = Thread.currentThread();
+    return true;
+  }
+
+  private @Nullable Tuple processStreamedResultsImpl(ProcessState state) throws SQLRuntimeException {
+    synchronized (this) {
```

Comment:
This likely goes away from the recent effort of removing `synchonized` for better Loom compatibility

Reply:
I didn't think we pushed that yet. I was going to put the loom stuff in an experimental branch until it's all sorted.